File: rust/auction-house/program/tests/update_auction_house.rs
// pub mod utils;

// #[cfg(test)]
// mod update_auction_house {

//     use super::utils::setup_functions::{setup_auction_house, setup_client, setup_program};
//     use anchor_client::{
//         solana_sdk::{signature::Keypair, signer::Signer, system_program, sysvar},
//         ClientError,
//     };
//     use mpl_auction_house::{
//         accounts as mpl_auction_house_accounts, instruction as mpl_auction_house_instruction,
//         AuctionHouse,
//     };

//     #[test]
//     fn success() -> Result<(), ClientError> {
//         // Payer Wallet
//         let payer_wallet = Keypair::new();

//         // Client
//         let client = setup_client(payer_wallet);

//         // Program
//         let program = setup_program(client);

//         // Airdrop the payer wallet
//         let signature = program
//             .rpc()
//             .request_airdrop(&program.payer(), 10_000_000_000)?;
//         program.rpc().poll_for_signature(&signature)?;

//         // Auction house authority
//         let authority_keypair = Keypair::new();
//         let authority = authority_keypair.pubkey();

//         // New auction house authority
//         let new_authority = Keypair::new().pubkey();

//         // Treasury mint key
//         let t_mint_key = spl_token::native_mint::id();

//         let auction_house_key = setup_auction_house(&program, &authority, &t_mint_key).unwrap();

//         let twd_key = program.payer();
//         let fwd_key = program.payer();
//         let tdw_ata = twd_key;

//         let seller_fee_basis_points: u16 = 345;

//         program
//             .request()
//             .signer(&authority_keypair)
//             .accounts(mpl_auction_house_accounts::UpdateAuctionHouse {
//                 treasury_mint: t_mint_key,
//                 payer: program.payer(),
//                 authority,
//                 new_authority,
//                 fee_withdrawal_destination: fwd_key,
//                 treasury_withdrawal_destination: tdw_ata,
//                 treasury_withdrawal_destination_owner: twd_key,
//                 auction_house: auction_house_key,
//                 token_program: spl_token::id(),
//                 system_program: system_program::id(),
//                 ata_program: spl_associated_token_account::id(),
//                 rent: sysvar::rent::id(),
//             })
//             .args(mpl_auction_house_instruction::UpdateAuctionHouse {
//                 seller_fee_basis_points: Some(seller_fee_basis_points),
//                 requires_sign_off: Some(false),
//                 can_change_sale_price: Some(false),
//             })
//             .send()?;

//         let auction_house_account: AuctionHouse = program.account(auction_house_key)?;

//         assert_eq!(
//             seller_fee_basis_points,
//             auction_house_account.seller_fee_basis_points
//         );

//         assert_eq!(false, auction_house_account.requires_sign_off);
//         assert_eq!(false, auction_house_account.can_change_sale_price);

//         Ok(())
//     }
// }

File: js/packages/web/src/actions/instructions/updateAuctionHouse.ts
import {
  PublicKey,
  PublicKeyInitData,
  TransactionInstruction,
} from '@solana/web3.js'
import { NATIVE_MINT } from '@solana/spl-token'
import { Wallet } from '@metaplex/js'
import { AuctionHouseProgram } from '@metaplex-foundation/mpl-auction-house'

const { createUpdateAuctionHouseInstruction } = AuctionHouseProgram.instructions

interface UpdateAuctionHouseParams {
  wallet: Wallet
  sellerFeeBasisPoints: number
  canChangeSalePrice?: boolean
  requiresSignOff?: boolean
  treasuryWithdrawalDestination?: PublicKeyInitData
  feeWithdrawalDestination?: PublicKeyInitData
  treasuryMint?: PublicKeyInitData
}

export const updateAuctionHouse = async (
  params: UpdateAuctionHouseParams
): Promise<TransactionInstruction> => {
  const {
    wallet,
    sellerFeeBasisPoints,
    canChangeSalePrice = false,
    requiresSignOff = false,
    treasuryWithdrawalDestination,
    feeWithdrawalDestination,
    treasuryMint,
  } = params

  const twdKey = treasuryWithdrawalDestination
    ? new PublicKey(treasuryWithdrawalDestination)
    : wallet.publicKey

  const fwdKey = feeWithdrawalDestination
    ? new PublicKey(feeWithdrawalDestination)
    : wallet.publicKey

  const tMintKey = treasuryMint ? new PublicKey(treasuryMint) : NATIVE_MINT

  const twdAta = tMintKey.equals(NATIVE_MINT)
    ? twdKey
    : (
        await AuctionHouseProgram.findAssociatedTokenAccountAddress(
          tMintKey,
          twdKey
        )
      )[0]

  const [auctionHouse] = await AuctionHouseProgram.findAuctionHouseAddress(
    wallet.publicKey,
    tMintKey
  )

  return createUpdateAuctionHouseInstruction(
    {
      treasuryMint: tMintKey,
      payer: wallet.publicKey,
      authority: wallet.publicKey,
      newAuthority: wallet.publicKey,
      feeWithdrawalDestination: fwdKey,
      treasuryWithdrawalDestination: twdAta,
      treasuryWithdrawalDestinationOwner: twdKey,
      auctionHouse,
    },
    {
      sellerFeeBasisPoints,
      requiresSignOff,
      canChangeSalePrice,
    }
  )
}
